### ISSUE TITLE
Json paths returned in rules need to be accurate to the last leaf level where the error occurs

### ISSUE THREAD

**Issue by veronicagg** [Azure/autorest#2106]
Here are some examples of rule violations, which are not accurate to the last level:
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L2664 
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L1269
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L1518
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L1734
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L2759
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L3609
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L3719
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L3741
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L4268
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/217096049#L5342

**Comment by veronicagg** [comment]
@sarangan12 I'm assigning this one to you, it involves many rules, so feel free to ask for help as needed. I opened only one issue to track this (at least for now) because it may make sense to update these paths in bulk. If you think it'd be better to split, feel free to do so.

**Comment by olydis** [comment]
Not sure if this is tracked somewhere already but it looks like the following rules report invalid source locations:

| Rule | Example of reported path |
|-----|-----|
| `TrackedResourceListByImmediateParent` | `$.paths["#/definitions/runs"]` (whoops) |
| `SecurityDefinitionsStructureValidation ` | `$.securityDefinitions` (can't blame that location if it doesn't exist) |
| `LongRunningResponseValidationRule ` | `$.<...>.extensions["x-ms-long-running-operation"]` |
| `NextLinkPropertyMustExist ` | `$.<...>.extensions["x-ms-pageable"]` |

The last two are a bit mean, i.e. "it's not your fault": The .NET part of AutoRest parses extensions (all stuff starting with `x-`) into a dictionary `extensions`, so the object tree looks different from the original Swagger! In other words, in the last example above, it should really have complained at `$.<...>["x-ms-pageable"]` because there *is no `extensions` property* in the Swagger. I'll look into fixing that.

**Comment by olydis** [comment]
The last two problems will be fixed with https://github.com/Azure/autorest/pull/2181.

**Comment by dsgouda** [comment]
@olydis there are 2 methods we use when implementing a rule, ```GetValidationMessages()``` and ```IsValid()```
It is possible to set the proper path in context with ```GetValidationMessages()``` but we can't do the same in ```IsValid()```, the linter logic automatically assigns the node to which the rule was applied as the appropriate path, which is fine until you run into something like ```SecurityDefinitionsStructureValidation``` where ServiceDefinition has an empty array even if the node itself is not explicitly declared in the actual spec.
tl;dr
Just make AutoRest core check if the node exists before reporting (if a node doesn't exist keep climbing up the path until you find a valid existing node), should be a clean fix even if we erroneously report the paths.

**Comment by olydis** [comment]
Sure, that would improve stability but I don't see how that is a fix. Tools report non-existent locations as the root of a problem (especially `TrackedResourceListByImmediateParent`, but that one also uses `GetValidationMessages` so it's fixable).

So I assume you're suggesting the fix in the core due to `SecurityDefinitionsStructureValidation`? Wouldn't it be cleaner to just make that use `GetValidationMessages`?

**Comment by dsgouda** [comment]
Yes, rules that need fixing will be fixed, I am on it. But the rules that use ```IsValid()``` don't need to be changed to ```GetValidationMessages()``` is what I think.

**Comment by dsgouda** [comment]
While I'm at it, trying to cleanup/improve some of the messages we display in our validation rules. Will post a list of rules where more information is required.

1. Rule M1005 (naming of GET operations) displays a message which reads like 
"'GET' operation 'Noun_NotNamedGet' must use method name 'Get' or Method name start with 'List'"
I believe this needs to be updated, any suggestions?

@kirthik ?
